signup/signup.js

```diff
@@ -13,6 +17,191 @@ $(document).ready(function() {
+			// called for every date before it is displayed
+			beforeShowDay: function(date) {
+				if (datesAllowed.includes(date.toDateString())) {
+					return [true, ''];
+				} else {
+					return [false, ''];
+				}
+			}
+		});
+
+		// Documentation: http://jonthornton.github.io/jquery-timepicker/
+		$("#timeInput").timepicker({
+			'step': 60,
+			'forceRoundTime': true,
+			'timeFormat': 'g:i A'
+		});
+	});
+}
+
+// This function will watch the site selector and the date selector for changes.
+function updateSitesDatesAndTimes() {
+	siteSelect = $("#sitePicker select");
+	dateInput = $("#dateInput");
+	// When the value of the site selector is updated...
+	// 1. Display the date picker
+	// 2. Update the options for the date selector to disable options that are not available at this site.
+	siteSelect.change(function() {
+		$("#datePicker").show();
+		// Clear any previously selected date
+		$("#dateInput").val('');
+		var value = this.value;
+		// Clear and repopulate the array of dates that are available at the selected site
+		datesAllowed.length = 0;
+		const i = sitesWithShifts.map(function(s) {return s.siteId}).indexOf(value);
+		if(i >= 0) {
+			datesAllowed = datesAllowed.concat(sitesWithShifts[i].dates.map(function(d) {return d.toDateString()}));
+			datesAllowed.sort(function(a,b) {
+				return Date.parse(a.start) - Date.parse(b.start);
+			});
+			$("#dateInput").datepicker("option", "minDate", new Date(datesAllowed[0]));
+			$("#dateInput").datepicker("option", "maxDate", new Date(datesAllowed[datesAllowed.length-1]));
+		}
+		// If a site is selected that has no available dates, show this to the user.
+		if(datesAllowed.length === 0) {
+			$("#dateInput").val('No dates available');
+			$("#dateInput").prop("disabled", true);
+			$("#timePicker").hide();
+		} else {
+			$("#dateInput").prop("disabled", false);
+		}
+	});
+
+	// When the date selector changes...
+	// 1. Display the time picker
+	// 2. Update the available times selector.
+	dateInput.change(function() {
+		$("#timePicker").show();
+		// Clear any previously selected time
+		$("#timeInput").val('');
+		var siteId = siteSelect.val();
+		var date = this.value;
+
+		let currentSite = sitesWithShifts.find(function(site) {return site.siteId === siteId});
+		if(currentSite !== undefined && currentSite.shiftTimes.length > 0) {
+			let availableTimes = [];
+			currentSite.shiftTimes.forEach(function(shiftTime) {
+				if($.datepicker.formatDate("mm/dd/yy",shiftTime.date) === date) {
+					availableTimes.push({'start':shiftTime.startTime, 'end':shiftTime.endTime})
+				}
+			});
+
+			if(availableTimes.length === 1) {
+				start = new Date(availableTimes[0].start);
+				end = new Date(availableTimes[0].end);
+				minTime = start.getHours() + ":" + start.getMinutes();
+				maxTime = end.getHours() + ":" + end.getMinutes();
+				$("#timeInput").timepicker('option', {'minTime': minTime, 'maxTime': maxTime});
+			} else if(availableTimes.length > 1) {
```

Comment:
What's the difference in logic between the ===1 case and the >1 case? They seem totally identical other than the sorting? And the sorting should work on an array of length 1, right? So couldn't these two cases be combined?